### ISSUE TITLE
MemoryError

### ISSUE THREAD

**Issue by bloomdt-uw** [yeatmanlab/pyAFQ#562]
The following `MemoryError` was encountered by Sendy, while attempting to segment a 7.79 Gb `tck` file from `qsiprep` reconstruction.

![image](https://user-images.githubusercontent.com/62235932/96503596-5af55d00-1208-11eb-9efa-845546ae9642.png)

**Comment by bloomdt-uw** [comment]
Maybe related to #564

**Comment by arokem** [comment]
I just realized that this is being thrown in DIPY after reading the
streamlines into memory. Speficially, during this call:
https://github.com/yeatmanlab/pyAFQ/blob/master/AFQ/segmentation.py#L524

So, in this case, there should be an opportunity to downsample from the 10M
down to a more reasonable number, before starting the segmentation process.
Alternatively, the `nb_points` keyword argument (
https://github.com/yeatmanlab/pyAFQ/blob/master/AFQ/segmentation.py#L67)
could be used to downsample each streamline to less points.

**Comment by bloomdt-uw** [comment]
Note that in addition to 32GB of RAM there is an additional 32GB of swap. In theory this should be more than sufficient to load a 7GB tck file.
![image](https://user-images.githubusercontent.com/62235932/96512386-453a6480-1215-11eb-9ef3-572582a93752.png)

**Comment by bloomdt-uw** [comment]
[sub-NLR160EK_ses-20160627_space-T1w_desc-preproc_space-T1w_desc-tracks_ifod2.tck](https://filebox.bcbl.eu/s/YNpAMin2XFLttyj)

**Comment by bloomdt-uw** [comment]
@arokem Tried locally over the weekend with and still got MemoryError

```
segmentation_params={
        "filter_by_endpoints": False,
        "nb_points": 50
    },
```

**Comment by bloomdt-uw** [comment]
See #565

**Comment by 36000** [comment]
So I think I know what is causing this. It's this line here:
https://github.com/yeatmanlab/pyAFQ/blob/2abfabf1c0268a7b03fc530c029af1bc14e2d4a1/AFQ/segmentation.py#L494
Normally the tractography is stored as a list of streamlines, where streamlines are of variable length. Here we turn the resampled streamlines into one giant array. We can do this because the resampled streamlines are no longer variable length. 

The problem with the giant array is that it needs a contiguous block of memory to allocate. In the example above, it is 10 million streamlines by 100 points by 3 4-byte floats, AKA 12 GB. However, it does not fail on that allocation; it is failing in the dipy code when it does `.astype(np.float)`. This turns it into a request for 24 contiguous GB, which causes the memory error.

TLDR; I think the solution is to remove np.array() from this line:
https://github.com/yeatmanlab/pyAFQ/blob/2abfabf1c0268a7b03fc530c029af1bc14e2d4a1/AFQ/segmentation.py#L494
I will make a PR tomorrow to try this out.

**Comment by richford** [comment]
Good find @bloomdt-uw! You might also want to consider using a dask array and calling `.compute()` when you need it later.
